alacritty/src/display/damage.rs

```diff
@@ -1,20 +1,197 @@
+        self.frames.swap(0, 1);
+    }
+
+    /// Resize the damage information in the tracker.
+    pub fn resize(&mut self, screen_lines: usize, columns: usize) {
+        self.screen_lines = screen_lines;
+        self.columns = columns;
+        for frame in &mut self.frames {
+            frame.reset(screen_lines, columns);
+        }
+        self.frame().damage_all = true;
+    }
+
+    /// Damage vi cursor inside the viewport.
+    pub fn damage_vi_cursor(&mut self, mut vi_cursor: Option<Point<usize>>) {
+        mem::swap(&mut self.old_vi_cursor, &mut vi_cursor);
+
+        if self.frame().damage_all {
+            return;
+        }
+
+        if let Some(vi_cursor) = self.old_vi_cursor {
+            self.frame().damage_point(vi_cursor);
+        }
+
+        if let Some(vi_cursor) = vi_cursor {
+            self.frame().damage_point(vi_cursor);
+        }
+    }
+
+    /// Get shaped frame damage for the active frame.
+    pub fn shape_frame_damage(&self, size_info: SizeInfo<u32>) -> Vec<Rect> {
+        if self.frames[0].damage_all {
+            vec![Rect::new(0, 0, size_info.width() as i32, size_info.height() as i32)]
+        } else {
+            let lines_damage = RenderDamageIterator::new(
+                TermDamageIterator::new(&self.frames[0].lines, 0),
+                &size_info,
+            );
+            lines_damage.chain(self.frames[0].rects.iter().copied()).collect()
+        }
+    }
+
+    /// Set the selection adding the damage for the current frame.
+    pub fn damage_selection(
+        &mut self,
+        mut selection: Option<SelectionRange>,
+        display_offset: usize,
+    ) {
+        mem::swap(&mut self.old_selection, &mut selection);
+
+        if self.frame().damage_all {
+            return;
+        }
+
+        if selection != self.old_selection {
+            for selection in self.old_selection.into_iter().chain(selection) {
+                // self.damage_selection_impl(selection, display_offset);
+                let display_offset = display_offset as i32;
+                let last_visible_line = self.screen_lines as i32 - 1;
+                let columns = self.columns;
+
+                // Don't damage invisible selection.
+                if selection.end.line.0 + display_offset < 0
+                    || selection.start.line.0.abs() < display_offset - last_visible_line
+                {
+                    continue;
+                };
+
+                let start = cmp::max(selection.start.line.0 + display_offset, 0);
+                let end = (selection.end.line.0 + display_offset).clamp(0, last_visible_line);
+                for line in start as usize..=end as usize {
+                    self.frame().lines[line].expand(0, columns - 1);
+                }
+            }
+        }
+    }
+}
+
+/// Damage state for the rendering frame.
+#[derive(Debug, Default)]
+pub struct FrameDamage {
+    /// The entire frame got redrawn.
+    damage_all: bool,
```

Comment:
```suggestion
    full: bool,
```

`damage_all` sounds more like a verb than an adjective. Should probably be `all`, though `full` sounds better in this context imo. `full_damage` would be another option if you want to be more verbose, but we're talking about `FrameDamage` already anyway so it shouldn't be necessary.